eth/connmanager.go

```diff
@@ -0,0 +1,254 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package eth
+
+import (
+	crand "crypto/rand"
+	"encoding/binary"
+	mrand "math/rand"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+)
+
+const (
+	// Interval between peer drop events (uniform between min and max)
+	peerDropIntervalMin = 3 * time.Minute
+	// Interval between peer drop events (uniform between min and max)
+	peerDropIntervalMax = 7 * time.Minute
+	// Avoid dropping peers for some time after connection
+	doNotDropBefore = 2 * peerDropIntervalMax
+	// How close to max should we initiate the drop timer. O should be fine,
+	// dropping when no more peers can be added. Larger numbers result in more
+	// aggressive drop behavior.
+	peerDropThreshold = 0
+	// Sync status poll interval (no need to be too reactive here)
+	syncCheckInterval = 60 * time.Second
+)
+
+// connManager monitors the state of the peer pool and makes changes as follows:
+//   - during sync the Downloader handles peer connections co connManager is disabled
+//   - if not syncing and the peer count is close to the limit, it drops peers
+//     randomly every peerDropInterval to make space for new peers
+//   - peers are dropped separately from the inboud pool and from the dialed pool
+type connManager struct {
+	connmanConfig
+	peersFunc   getPeersFunc
+	syncingFunc getSyncingFunc
+
+	// The peerDrop timers introduce churn if we are close to limit capacity.
+	// We handle Dialed and Inbound connections separately
+	peerDropDialedTimer  *mclock.Alarm
+	peerDropInboundTimer *mclock.Alarm
+
+	peerEventCh chan *p2p.PeerEvent // channel for peer event changes
+	sub         event.Subscription  // subscription to peerEventCh
+
+	wg         sync.WaitGroup // wg for graceful shutdown
+	shutdownCh chan struct{}
+}
+
+// Callback type to get the list of connected peers.
+type getPeersFunc func() []*p2p.Peer
+
+// Callback type to get syncing status.
+// Returns true while syncing, false when synced.
+type getSyncingFunc func() bool
+
+type connmanConfig struct {
+	maxDialPeers    int // maximum number of dialed peers
+	maxInboundPeers int // maximum number of inbound peers
+	log             log.Logger
+	clock           mclock.Clock
+	rand            *mrand.Rand
+}
+
+func (cfg connmanConfig) withDefaults() connmanConfig {
+	if cfg.log == nil {
+		cfg.log = log.Root()
+	}
+	if cfg.clock == nil {
+		cfg.clock = mclock.System{}
+	}
+	if cfg.rand == nil {
+		seedb := make([]byte, 8)
+		crand.Read(seedb)
+		seed := int64(binary.BigEndian.Uint64(seedb))
+		cfg.rand = mrand.New(mrand.NewSource(seed))
+	}
+	return cfg
+}
+
+func newConnManager(config *connmanConfig) *connManager {
```

Comment:
To subscribe to syncing, pass `event.TypeMux` pointer here from [`Ethereum.eventMux`](https://github.com/ethereum/go-ethereum/blob/9f83e9e6734f82facbc66fd4cd591dd648aaec92/eth/backend.go#L83C2-L83C10)

and subscribe to events using 

```
cm.syncsub = eventMux.Subscribe(downloader.StartEvent{}, downloader.StopEvent{})
```

Reply:
Actually I don't need to subscribe anymore, as I just need the current status. Also, if the status just filpped, it won't make a difference to read the wrong value for this use case.